### PR TITLE
Configuration check helpers

### PR DESCRIPTION
**Pull request by bobdenotter**
Adds a configuration notice when running "debug" in PROD. Comments welcome. 

screen shot 2016-12-04 at 14 28 52

### CODE REVIEW

File: src/Config.php [hunk 1]
```diff
@@ -371,6 +371,12 @@ protected function parseGeneral()
             $general['branding']['provided_by']
         );
 
+        // Add some common 'development' hosts to the `debug_local_domains` setting:
+        $general['debug_local_domains'] = array_unique(array_merge(
+            (array) $general['debug_local_domains'],
+            ['.dev', 'dev.', 'devel.', 'development.', 'test.', '.test', 'new.', '.new', 'localhost', '.local', 'local.']
+        ));
+
```
Comment (GawainLynch) [inline src/Config.php]: And if you're production domain legitimately uses one of these?

File: src/Config.php [hunk 2]
```diff
@@ -1128,6 +1134,7 @@ protected function getDefaults()
             'debug_enable_whoops'         => false, /** @deprecated. Deprecated since 3.2, to be removed in 4.0 */
             'debug_error_use_symfony'     => false,
             'debug_permission_audit_mode' => false,
+            'debug_local_domains'         => [],
```
Comment (GawainLynch) [inline src/Config.php]: So the people that can't remember to run it off, are going to remember to configure this? Yeah, nah!

File: src/Config.php (hunk 1 above)
Comment (bobdenotter) [inline src/Config.php]: Doesn't matter. In the off chance your domain is `locahost.org`, you just won't see the notice, no harm done.

File: src/Config.php (hunk 2 above)
Comment (bobdenotter) [inline src/Config.php]: No, the opposite of that. If you want to get rid of the message, add your dev domain here.

**Comment by bobdenotter** [comment]
I consolidated the 'configuration notices' and the 'development version' warning into one. 

screen shot 2016-12-05 at 14 34 52

**Comment by bobdenotter** [comment]
Rebased and all. Should be GTG now.

**Comment by bobdenotter** [comment]
Green lights. 🚦

**Comment by CarsonF** [comment]
I'm on the fence about this. But the first thing that comes to mind is that this is a check for novices, and novices aren't going to add their "dev domains" to the config. Thus it doesn't even need to be configurable. 

I would rather have these things as a separate composer package, that's included by default with `composer-install`. But it's easy to remove for those that don't want it.

If we insist on keeping these things in core, I would like a "I know what I'm doing" config option.

**Comment by GawainLynch** [comment]
> I would rather have these things as a separate composer package

Ha! You took the words right out of my mouth. I was just thinking though this and came to the same conclusion. 

An even better option would be to have said Composer package install the archive version, that way people doing a default `composer-install` for their base builds can not worry either.

**Comment by bobdenotter** [comment]
> But the first thing that comes to mind is that this is a check for novices, and novices aren't going to add their "dev domains" to the config. Thus it doesn't even need to be configurable.

I don't think that's the point of this, at all. This is something for novices, agreed, but they don't need to add their "dev domains" to config. They _only_ need to do this if they've set up a config name that does _not_ have `.local` or `.dev` or one of the others in it. If they did do that, I wouldn't consider them novices per se. 

In my mind this check _does_ have edge cases where it doesn't "trigger" on production sites. Sure, but the worst that can happen is that it won't show the notice for the dev domains, and it'll act like it does now. In the other 99% of the cases it _will_ work, and it'll help those novices by urging them to turn off debug in prod.  

> I would rather have these things as a separate composer package

Sure, we can add them as composer packages later. Gawain mentioned today that he'd like to move all of these checks out of `GeneralListener.php`. But, that'll take work, and this PR - as it stands - is an improvement over that we have right now.

**Comment by CarsonF** [comment]
Yeah I'm fine with that being the next step. Either way we both agree to remove the config option.

**Comment by bobdenotter** [comment]
> Either way we both agree to remove the config option.

What? no. 

If we do that, Gawain will _always_ see the notice, when he's developing. I'm sure he'll get annoyed by that real quickly.

**Comment by CarsonF** [comment]
Oh I see. The config option is there to disable the check. Why not just have a `check_dev_domain` boolean option instead then?

File: src/EventListener/ConfigListener.php [hunk 3]
```diff
@@ -161,6 +161,8 @@ private function setGlobals()
             $usersCur = null;
         }
 
+        $twig->addGlobal('bolt_name', Version::name());
+        $twig->addGlobal('bolt_version', Version::VERSION);
```
Comment (bobdenotter) [inline src/EventListener/ConfigListener.php]: Those were already there.

Comment (bobdenotter) [inline src/EventListener/ConfigListener.php]: Ah, I see.. They snuck in from here.. https://github.com/bolt/bolt/blob/release/3.2/src/EventListener/ConfigListener.php#L165-L167

I'll remove them.

**Comment by bobdenotter** [comment]
> Why not just have a check_dev_domain boolean option instead then?

Let's say your dev environment is `foo.bar`, and your site will be hosted on `baz.qux`. You start devving, and you get annoyed about the message popping up, as you should.. You then set `debug_local_domains:  [.bar ]`, so it stops bugging you. 

You move the site to PROD at http://baz.qux, forget to turn off debug, and hand over the keys to your editor. The editor will then ask you "So, what is this annoying message about debug in the backend?". You then turn off debug in `config_local.yml`, and all is well. 

president-george-w -bush-mission-accomplished

**Comment by CarsonF** [comment]
> forget to turn off debug

I was assuming that if you disabled the check that doesn't happen.

`debug` is based on an environment variable for me, I don't even use config_local.yml.

**Comment by CarsonF** [comment]
Also `bolt_stable` twig global was there as a way to disable that notice. So if we are removing that then we need to add a config option to disable that check as well.

**Comment by bobdenotter** [comment]
> I was assuming that if you disabled the check that doesn't happen. And if it does that's on you. 

That is the entire point. This is not for you. This is for (mainly) developers who don't have an ingrained mental connect between "debug is on" and "my cache will get out of control in prod if i don't fix that". 

It's a small nudge for those devs to turn it off, should they forget to do so. Or for those who don't even realise that they should do it.

**Comment by bobdenotter** [comment]
> Also bolt_stable twig global was there as a way to disable that notice.

That is something entirely different, mate. I put that in this PR as suggested by Gawain. Basically this PR _consolidates_ the different configuration checks (mail, the 'check for debug on prod' and the 'running unstable versions') into _one_ sidebar widget, instead of spreading them out into multiple.

**Comment by CarsonF** [comment]
How about this, since we can't move it to separate package yet, can you move those to a `NoviceListener`? That way we can just disable it that way. No need to go through the config hassle.

**Comment by bobdenotter** [comment]
> those

Which "those", exactly? all of them?

**Comment by bobdenotter** [comment]
Also, why would you want to disable them?

**Comment by GawainLynch** [comment]
So … 5 minutes https://github.com/GawainLynch/starter-kit

Next 5 minutes the rest of it …

**Comment by CarsonF** [comment]
all of them. I don't need them and I don't want them. I'm running 3.3 on a production site right now. I don't want my editors to see that notice. Is that advised? Of course not. But it's my call as the system owner, not Bolt's.

**Comment by bobdenotter** [comment]
> I'm running 3.3 on a production site right now. I don't want my editors to see that notice.

You really, really are missing the point.. _**YOU**_ won't see this message, at all! If you have your stuff set up correctly, you, nor your editors will see it. EVER.

**Comment by CarsonF** [comment]
The domain check, yes. But there's no way to disable the dev version check now.

**Comment by bobdenotter** [comment]
> But there's no way to disable the dev version check now.

No, but there was no way to turn that off either before this PR.

**Comment by CarsonF** [comment]
```php
// Disable development warning.
$this->on(KernelEvents::REQUEST, function () {
    $this['twig']->addGlobal('bolt_stable', true);
}, 30);
```

**Comment by CarsonF** [comment]
It's hacky, but doable. I would appreciate a better way to do it.

**Comment by bobdenotter** [comment]
I just moved it _out_ of twig and into code, because I thought we all agreed there should be less logic in code. :'-(

**Comment by CarsonF** [comment]
> I just moved it out of twig and into code, because I thought we all agreed there should be less logic in code. :'-(

Definitely. The way I was doing it is wrong. But it needs to be configurable.

I am just looking at the rest of `GeneralListener`. I don't need any of it, so I can just disable that class and I'll be fine.

**Comment by bobdenotter** [comment]
Well, if we can make it a separate package that's included by default (like `bolt/thumbs`), i can live with that. @GawainLynch's repo looks like a good approach to me.

**Comment by GawainLynch** [comment]
OK, Done … If we agree, we'll just register `bolt/starter-kit` on Packagist.

… and I have moved the repo to the org https://github.com/bolt/starter-kit

**Comment by CarsonF** [comment]
As long as bolt doesn't depend on that package it looks good to me.

**Comment by GawainLynch** [comment]
We can then just have something like:
```
        if (class_exists('\Bolt\Starter\Provider\StarterProvider')) {
            $this->register(new \Bolt\Starter\Provider\StarterProvider());
        }
```

**Comment by GawainLynch** [comment]
> As long as bolt doesn't depend on that package it looks good to me.

Archive distribution installation addition.

**Comment by bobdenotter** [comment]
> $this->register(new \Bolt\Starter\Provider\StarterProvider()); 

In `Application.php` or somesuch, right? 

> Archive distribution installation addition.

So, we'll add `bolt/starter-kit` to the composer.json here: https://github.com/bolt/composer-install/blob/release/3.2/composer.json ?

This looks like a good solution to me. I'll be able to add some more "end user friendly" checks there, then. :-)

**Comment by GawainLynch** [comment]
> So, we'll add bolt/starter-kit to the composer.json here: https://github.com/bolt/composer-install/blob/release/3.2/composer.json ?

No, just ask part of the script that builds the distribution. it can do a `composer require` and will just add it to the root project, making it removable afterwards.

**Comment by GawainLynch** [comment]
Packagist + webhooks all set up, so`composer require bolt/starter-kit:^1.0@dev` should now work

>> $this->register(new \Bolt\Starter\Provider\StarterProvider());

>In Application.php or somesuch, right?

Yeah, bottom of the providers section I would guess to be best for now.

**Comment by GawainLynch** [comment]
> This looks like a good solution to me. I'll be able to add some more "end user friendly" checks there, then. :-)

Also what I was thinking. I honestly have no objections to having installs that are easier for certain groups, the easier the better. Just as long as we have choice :+1:

**Comment by bobdenotter** [comment]
> No, just ask part of the script that builds the distribution. it can do a composer require and will just add it to the root project,

I'm not seeing how to implement that properly yet. 

> Yeah, bottom of the providers section I would guess to be best for now.

Ok.

**Comment by GawainLynch** [comment]
After this https://github.com/bolt/bolt-distribution/blob/master/package.sh#L36

You add:
```
composer require bolt/starter-kit:^1.0@dev --working-dir=$COMPILE_DIR
```

**Comment by bobdenotter** [comment]
One leftover tidbit.. What are we gonna do with `debug_local_domains: [ ] ` in `config.yml.dist` ?

**Comment by GawainLynch** [comment]
> One leftover tidbit.. What are we gonna do with debug_local_domains: [ ] in `config.yml.dist` ?

My thought is to add it to the docs only, and have sensible defaults in the package itself. The plan is to reduce what is in  `config.yml.dist` by default, and it is already an intimidating wall to the novice user 

… even this not so novice user gets annoyed at the default size :laughing:

**Comment by bobdenotter** [comment]
I'm having trouble getting this to work. How can I get a setup with this running?

**Comment by bobdenotter** [comment]
It looks like this now: 

screen shot 2016-12-06 at 17 57 16

this depends on the PR to `bolt/starter-kit` being merged, oc.

**Comment by bobdenotter** [comment]
See https://github.com/bolt/starter-kit/pull/1/files

**Comment by GawainLynch** [comment]
I am really liking where we went with this … it bridges gaps :+1:

**Pull request merged**